### ISSUE TITLE
making .cuda() falls back to an identity function when gpu is not available

### ISSUE THREAD

**Issue by kyunghyuncho** [pytorch/pytorch#3600]
it would be really convenient, if .cuda() could automatically fall back to doing nothing when there's no gpu available on the machine. it was suggested by @khanhptnk

**Comment by khanhptnk** [comment]
Thanks @kyunghyuncho :)

**Comment by kyunghyuncho** [comment]
would this be resolved by modifying ``CudaTransfer``?

https://github.com/pytorch/pytorch/blob/fde355f7d4ad95a1f8cba6adc3fff7b4935ecba0/torch/autograd/_functions/tensor.py#L172-L190

**Comment by kyunghyuncho** [comment]
maybe check ``torch.cuda.is_available()``, and if False, return i as it is?

**Comment by fmassa** [comment]
I'm not sure this is a good strategy, as it might mask from the user configuration problems that he might not be aware of.
I think the best would be to simplify the task of having device-agnostic code, so that instead of calling `.cuda()` one instead do something like `.type(my_type)`.
Adding support for 'dtype' in the constructor of some functions will be a first step towards that direction.

**Comment by kyunghyuncho** [comment]
the trouble i see with device-agnostic code is that one might want to mix in variables on both gpu and cpu. wouldn't it be easier to choose a default device to be cpu and let cuda fall back to doing nothing? in this case, if the code was intentionally implemented to mix variables on gpu and cpu, it would do so as intended with gpu is available, and otherwise, both types of variables would be on cpu.

using ``.type()`` sounds reasonable, but i can't think of how to make it automatic for all the different tensor types..

**Comment by fmassa** [comment]
I'm not sure having a fallback of "`gpu` doing nothing" is a good thing. I think it's better to say that it's unsupported than do things behind your back. Mixing CPU and GPU code makes writing device-agnostic code harder indeed, but that might be the price to pay to have such flexibility.

Another possibility is to perform automatic device conversion, with a default being specified by the user (so `op(cpu_tensor, gpu_tensor)` can be either run as moving everything to the cpu or to the gpu). The drawback is that it could become very easy to write non-performant code, as the CPU-GPU copy is expensive.

But that's just my 2 cents, I'll let the others comment as well. :)

**Comment by fmassa** [comment]
Also, see https://github.com/pytorch/pytorch/issues/3202 for other proposals for device-agnostic code

**Comment by khanhptnk** [comment]
I really like the idea about adding dtype to the Tensor constructor [#3202](https://github.com/pytorch/pytorch/issues/3202). Or similarly, you can add a is_cuda boolean to the constructor. My main irritant now is that I have to make the same if statement to check for cuda availability every time I need to create a new Tensor. So we can "cache" the result of that if statement into a boolean variable and just pass it to some function that will execute that if statement under the hood. We don't really change any behavior but now the if statement is only written once. 

Looking forward to this change!

**Comment by fmassa** [comment]
Note that for tensor construction, you can do something like
```python
tensor_type = torch.cuda.FloatTensor

my_tensor = tensor_type(2, 3)
my_tensor2 = torch.rand(2).type(tensor_type)
```

but having the `dtype` argument will be much better.

**Comment by hmc-cs-mdrissi** [comment]
If doing nothing is bad, then instead you could have it print a warning. Or alternatively you could add an argument to .cuda like fail_silently that defaults to False, but can be made true if the user wants it to check if cuda is available before trying to do anything. Equivalently a function called .try_cuda would make that intent pretty clear in the code.

**Comment by kyunghyuncho** [comment]
how about then printing a warning?

**Comment by elbaro** [comment]
+1 for warning and fallback.

meanwhile this is a simple hack:

```
def cuda(x, **kwargs):
    if config.gpu:
        return x.cuda(**kwargs)
    return x

var = Variable(cuda(tensor))
net = cuda(Network())
```

**Comment by yjchun** [comment]
+1 for warning and fallback! 
I am switching from keras to pytorch and found cuda check everywhere is very annoying and makes my code even uglier.

**Comment by gwgundersen** [comment]
I like the idea of a warning and fallback as well. But FWIW, I currently just use this utility function:

```python
def cudaize(obj):
    return obj.cuda() if torch.cuda.is_available() else obj
```

**Comment by carlosgmartin** [comment]
I second printing a warning (which can be turned off) if cuda is not available, or any of the other suggestions @hmc-cs-mdrissi proposed. Makes code cleaner with fewer if statements all over the place.
